File: inst/python/tests/test_validate_docx.py
import shutil
import tempfile
from pathlib import Path

from docx import Document

from reportipyr.validate import validate_docx

DATA_DIR = Path(__file__).parent / "data"


def _make_docx_with_text(text: str) -> str:
    path = tempfile.NamedTemporaryFile(suffix=".docx", delete=False).name
    doc = Document()
    doc.add_paragraph(text)
    doc.save(path)
    return path


def test_validate_docx_corrupt_file():
    path = tempfile.NamedTemporaryFile(suffix=".docx", delete=False).name
    Path(path).write_text("this is not a docx file")
    result = validate_docx(path, strict=True)
    assert result["success"] is False
    assert any("Failed to read document" in msg for msg in result["errors"])


def test_validate_docx_missing_magic():
    path = _make_docx_with_text("no magic here")
    result = validate_docx(path, strict=True)
    assert result["success"] is True
    assert any("does not contain magic strings" in msg for msg in result["warnings"])
    assert result["errors"] == []
    assert result["file_names"] == []


def test_validate_docx_invalid_extension_strict():
    path = _make_docx_with_text("{rpfy}:example.doc")
    result = validate_docx(path, strict=True)
    assert result["success"] is False
    assert any("Unsupported file types" in msg for msg in result["errors"])


def test_validate_docx_invalid_extension_nonstrict():
    path = _make_docx_with_text("{rpfy}:example.doc")
    result = validate_docx(path, strict=False)
    assert result["success"] is True
    assert any("Unsupported file types" in msg for msg in result["warnings"])


def test_validate_docx_duplicates_strict():
    path = _make_docx_with_text("{rpfy}:[a.csv, a.csv]")
    result = validate_docx(path, strict=True)
    assert result["success"] is False
    assert any("Found duplicate files" in msg for msg in result["errors"])


def test_validate_docx_duplicates_nonstrict():
    path = _make_docx_with_text("{rpfy}:[a.csv, a.csv]")
    result = validate_docx(path, strict=False)
    assert result["success"] is True
    assert any("Found duplicate files" in msg for msg in result["warnings"])


def test_validate_docx_shell_strict_passes():
    source_doc = DATA_DIR / "test-doc-shell.docx"
    docx_in = tempfile.NamedTemporaryFile(suffix=".docx", delete=False).name
    shutil.copyfile(source_doc, docx_in)

    result = validate_docx(docx_in, strict=True)
    assert result["success"] is True
    assert result["errors"] == []


def test_validate_docx_shell_nonstrict_passes():
    source_doc = DATA_DIR / "test-doc-shell.docx"
    docx_in = tempfile.NamedTemporaryFile(suffix=".docx", delete=False).name
    shutil.copyfile(source_doc, docx_in)

    result = validate_docx(docx_in, strict=False)
    assert result["success"] is True

File: tests/testthat/test-validate_docx.R
create_docx_with_magic_string <- function(magic_string) {
  path <- tempfile(fileext = ".docx")
  doc <- officer::read_docx()
  doc <- officer::body_add_par(doc, magic_string)
  print(doc, target = path)
  withr::defer(unlink(path), envir = parent.frame())
  path
}

create_config_yaml <- function(strict = TRUE) {
  path <- tempfile(fileext = ".yaml")
  writeLines(paste0("strict: ", tolower(strict)), path)
  withr::defer(unlink(path), envir = parent.frame())
  path
}

mock_paths <- list(venv = "/fake/.venv", uv = "/fake/uv")

test_that("validate_docx succeeds with valid docx and .csv file", {
  docx <- create_docx_with_magic_string("{rpfy}:example.csv")
  config <- create_config_yaml(strict = TRUE)

  mockery::stub(validate_docx, "pyro::get_venv_uv_paths", function() mock_paths)
  mockery::stub(validate_docx, "processx::run", function(...) {
    list(stdout = jsonlite::toJSON(list(
      success = TRUE,
      file_names = list("example.csv"),
      warnings = list(),
      errors = list()
    )))
  })

  expect_silent(validate_docx(docx, config))
})

test_that("validate_docx errors if file extension is invalid", {
  docx <- create_docx_with_magic_string("{rpfy}:example.doc")
  config <- create_config_yaml(strict = TRUE)

  mockery::stub(validate_docx, "pyro::get_venv_uv_paths", function() mock_paths)
  mockery::stub(validate_docx, "processx::run", function(...) {
    list(stdout = jsonlite::toJSON(list(
      success = FALSE,
      file_names = list("example.doc"),
      warnings = list(),
      errors = list(
        "Unsupported file types found in document: example.doc",
        "Fix artifact extensions to continue. Currently .csv, .RDS are accepted for tables and .png is accepted for figures."
      )
    )))
  })

  expect_error(validate_docx(docx, config), "Fix artifact extensions")
})

test_that("validate_docx errors on duplicated files in strict mode", {
  docx <- create_docx_with_magic_string("{rpfy}:[example.csv, example.csv]")
  config <- create_config_yaml(strict = TRUE)

  mockery::stub(validate_docx, "pyro::get_venv_uv_paths", function() mock_paths)
  mockery::stub(validate_docx, "processx::run", function(...) {
    list(stdout = jsonlite::toJSON(list(
      success = FALSE,
      file_names = list("example.csv", "example.csv"),
      warnings = list(),
      errors = list(
        "Found duplicate files, please fix: example.csv",
        "Using strict mode. Fix duplicate artifacts to continue."
      )
    )))
  })

  expect_error(
    validate_docx(docx, config),
    "Using strict mode. Fix duplicate artifacts to continue."
  )
})

test_that("validate_docx errors if input .docx does not exist", {
  fake_path <- tempfile(fileext = ".docx")

  expect_error(
    validate_docx(fake_path, NULL),
    "The input document does not exist"
  )
})

test_that("validate_docx warns and succeeds if magic strings are missing", {
  docx <- tempfile(fileext = ".docx")
  print(officer::read_docx(), target = docx)
  config <- create_config_yaml()

  mockery::stub(validate_docx, "pyro::get_venv_uv_paths", function() mock_paths)
  mockery::stub(validate_docx, "processx::run", function(...) {
    list(stdout = jsonlite::toJSON(list(
      success = TRUE,
      file_names = list(),
      warnings = list("The file does not contain magic strings; nothing to process."),
      errors = list()
    )))
  })

  expect_no_error(validate_docx(docx, config))
})

test_that("validate_docx errors if .venv directory is missing", {
  docx <- create_docx_with_magic_string("{rpfy}:example.csv")

  mockery::stub(validate_docx, "pyro::get_venv_uv_paths", function() {
    stop("Create virtual environment with initialize_python")
  })

  expect_error(
    validate_docx(docx, NULL),
    "Create virtual environment with initialize_python"
  )
})

test_that("validate_docx errors if uv path is NULL", {
  docx <- create_docx_with_magic_string("{rpfy}:example.csv")

  mockery::stub(validate_docx, "pyro::get_venv_uv_paths", function() {
    stop("Please install uv with initialize_python")
  })

  expect_error(
    validate_docx(docx, NULL),
    "Please install uv with initialize_python"
  )
})
